src/runtime.rs

```diff
@@ -254,29 +254,95 @@ pub struct Parser<'i, P: ParseNodeKind, C: CodeLabel, I: Input> {
     pub sppf: ParseForest<'i, P, I>,
 }
 
-impl<'i, P: ParseNodeKind, C: CodeLabel, I: Input> Parser<'i, P, C, I> {
-    pub fn with<R>(input: I, f: impl for<'i2> FnOnce(Parser<'i2, P, C, I>, Range<'i2>) -> R) -> R {
+#[derive(Debug)]
+pub struct ParseError<T> {
+    pub partial: Option<T>,
+}
+
+impl<T> ParseError<T> {
+    pub fn map_partial<U>(self, f: impl FnOnce(T) -> U) -> ParseError<U> {
+        let ParseError { partial } = self;
+        ParseError {
+            partial: partial.map(f),
+        }
+    }
+
+    // FIXME(eddyb) remove this when `generate::rust` stops using it
+    pub fn as_ref_partial<'a>(&'a self) -> ParseError<&'a T> {
+        let ParseError { partial } = self;
+        ParseError {
+            partial: partial.as_ref(),
+        }
+    }
+}
+
+pub type ParseResult<T> = Result<T, ParseError<T>>;
+
+impl<'i, P: ParseNodeKind, C: CodeStep<P, I>, I: Input> Parser<'i, P, C, I> {
+    pub fn parse_with<R>(
+        input: I,
+        callee: C,
+        kind: P,
+        f: impl for<'i2> FnOnce(ParseResult<(ParseForest<'i2, P, I>, ParseNode<'i2, P>)>) -> R,
+    ) -> R {
         scope(input.to_container(), |input| {
-            let range = input.range();
-            f(
-                Parser {
-                    threads: Threads {
-                        queue: BinaryHeap::new(),
-                        seen: BTreeSet::new(),
-                    },
-                    gss: GraphStack {
-                        returns: HashMap::new(),
-                    },
-                    memoizer: Memoizer {
-                        lengths: HashMap::new(),
-                    },
-                    sppf: ParseForest {
-                        input,
-                        possibilities: HashMap::new(),
-                    },
+            let call = Call {
+                callee,
+                range: Range(input.range()),
+            };
+            let mut parser = Parser {
+                threads: Threads {
+                    queue: BinaryHeap::new(),
```

Comment:
Use `Default` here and below?

Reply:
I think the intention was that these types were not constructible from the outside.